File: 0238-product-of-array-except-self/solution.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] l = new int[n];
        int[] r = new int[n];
        l[0] = nums[0];
        r[n-1] = nums[n-1];
        for(int i=1;i<n;i++){
            l[i] = l[i-1]*nums[i];
            r[n-i-1] = r[n-i]*nums[n-i-1];
        }
        for(int i=0;i<n;i++){
            int x=1,y=1;
            if(i!=0)
            x = l[i-1];
            if(i!=n-1)
            y = r[i+1];
            nums[i] = x*y;
        }
        return nums;
    }
}

File: 0238-product-of-array-except-self/solution.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        pre = [1]
        suf = [1]
        # nums = [1,2,3,4]
        # pre = [1,1,2,6]
        # suf = [24,12,4,1]
        for i in range(1,n):
            pre.append(pre[-1]*nums[i-1])
        for i in range(n-1,0,-1):
            suf.insert(0,suf[0]*nums[i])
        res = []
        for i in range(n):
            res.append(pre[i]*suf[i])
        return res
